### ISSUE TITLE
error: Program is not supported by this target, because table .iFwdBrgMacKeyTbl has multiple successors 

### ISSUE THREAD

**Issue by thinkoco** [p4lang/p4c#1394]
Can I access same table twice in p4 code? I want to do lookup at same table with two different keys.So  my code like this:
```
table iFwdBrgMacKeyTbl {
        key = {
            keyOfiFwdBrgMacKeyTbl: exact;
        }
        actions= {'''}
        size = xxx;
}

apply{
        if (condition A){
            keyOfiFwdBrgMacKeyTbl = ABC;
            iFwdBrgMacKeyTbl.apply();
         }else{
            keyOfiFwdBrgMacKeyTbl = XYZ;
            iFwdBrgMacKeyTbl.apply();
         }
}
```
Is there any way that I can access same table in data plane twice? Any help will be appreciated.

**Comment by ChrisDodd** [comment]
The obvious way is to only apply the table once.

        if (condition A) {
            keyOfiFwdBrgMacKeyTbl = ABC;
         } else {
            keyOfiFwdBrgMacKeyTbl = XYZ;
         }
         iFwdBrgMacKeyTbl.apply();

**Comment by thinkoco** [comment]
@ChrisDodd Ths.Yes,ABC or XYZ is OK. Now, I am trying to find a  way to do  lookup with both  s-mac and d-mac.

**Comment by SantiagoBautista** [comment]
My suggestion would be to duplicate your table: have two tables with the same content but different names&key.
Other options include re-circulating the packet.
On Jul 9, 2018 10:42 PM, Welon Tank <notifications@github.com> wrote:Can I access same table twice in p4 code? I want to do lookup at same table with two different keys.So  my code like this:
table iFwdBrgMacKeyTbl {
        key = {
            keyOfiFwdBrgMacKeyTbl: exact;
        }
        actions= {'''}
        size = xxx;
}

apply{
        if (condition A){
            keyOfiFwdBrgMacKeyTbl = ABC;
            iFwdBrgMacKeyTbl.apply();
         }else{
            keyOfiFwdBrgMacKeyTbl = XYZ;
            iFwdBrgMacKeyTbl.apply();
         }
}

Is there any way that I can access same table in data plane twice? Any help will be appreciated.

**Comment by ChrisDodd** [comment]
P4 is deliberately designed with no loops, to model a "once through" pipeline for packets -- each table sees each packet, doing a single lookup, and then goes on to the next packet.  So if you want two lookups, you'll need to either use two tables, or arrange to process the entire packet twice (with resubmit or recirculate)

**Comment by jnfoster** [comment]
@ChrisDodd Actually, there is nothing in the P4 language that mandates at-most-once execution for each table, although many targets do.

**Comment by ChrisDodd** [comment]
Yes, there's a tension between the language design as being as general purpose as practical, against the target domain of packet processing and forwarding.  I think it's safe to say that all targets (currently and anticipated) require at most once execution of tables.

**Comment by jnfoster** [comment]
There are targets that support multiple executions of a table.

**Comment by mbudiu-vmw** [comment]
The ebpf backend supports multiple table reads. It's a limitation of each target, not the language or compiler.
